### ISSUE TITLE
Potential bug in mask propagation

### ISSUE THREAD

**Issue by AakashKumarNain** [keras-team/keras#18543]
I noticed a strange (potential) bug yesterday when developing a simple model. The model works fine with `tf.keras` but in Keras 3, it throws a warning which appears to be problematic. Here is the code to reproduce it:

## tf.keras

```python

class PositionalEmbedding(layers.Layer):
    def __init__(self, seq_length, input_dim, output_dim, **kwargs):
        super().__init__(**kwargs)
        self.seq_length = seq_length
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.token_embeddings = layers.Embedding(input_dim=input_dim, output_dim=output_dim)
        self.position_embeddings = layers.Embedding(input_dim=seq_length, output_dim=output_dim)
        self.supports_masking = True
        
    def call(self, inputs):
        seq_length = tf.shape(inputs)[-1]
        positions = tf.range(start=0, limit=seq_length, delta=1)
        embedded_tokens = self.token_embeddings(inputs)
        embedded_positions = self.position_embeddings(positions)
        return embedded_tokens + embedded_positions

    def compute_mask(self, inputs, mask=None):
        return tf.math.not_equal(inputs, 0)

    def get_config(self):
        config = super().get_config()
        config.update({
            "seq_length": self.sequence_length,
            "input_dim" : self.input_dim,
            "output_dim": self.output_dim,
        })
        return config

vocab_size = 10000
sequence_length = 200
embed_dim = 256
num_heads = 2
dense_dim = 32

inputs = keras.Input(shape=(None,), dtype="int64")
x = PositionalEmbedding(sequence_length, vocab_size, embed_dim)(inputs)
x = layers.GlobalMaxPooling1D()(x)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(1, activation="sigmoid")(x)
model = keras.Model(inputs, outputs)
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.summary()
```

<img width="696" alt="Screenshot 2023-10-04 at 10 24 41 AM" src="https://github.com/keras-team/keras/assets/11736571/108b6bc3-bae5-4f1d-99b2-fd885dd767bf">

---

## Keras 3

```python
class PositionalEmbedding(layers.Layer):
    def __init__(self, seq_length, input_dim, output_dim, **kwargs):
        super().__init__(**kwargs)
        self.seq_length = seq_length
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.token_embeddings = layers.Embedding(input_dim=input_dim, output_dim=output_dim)
        self.position_embeddings = layers.Embedding(input_dim=seq_length, output_dim=output_dim)
        self.supports_masking = True
        
    def call(self, inputs):
        seq_length = ops.shape(inputs)[-1]
        positions = jnp.arange(start=0, stop=seq_length, step=1)
        embedded_tokens = self.token_embeddings(inputs)
        embedded_positions = self.position_embeddings(positions)
        return embedded_tokens + embedded_positions

    def compute_mask(self, inputs, mask=None):
        return ops.not_equal(inputs, 0)

    def get_config(self):
        config = super().get_config()
        config.update({
            "seq_length": self.sequence_length,
            "input_dim" : self.input_dim,
            "output_dim": self.output_dim,
        })
        return config

vocab_size = 10000
sequence_length = 200
embed_dim = 256
num_heads = 2
dense_dim = 32

inputs = keras.Input(shape=(None,), dtype="int64")
x = PositionalEmbedding(sequence_length, vocab_size, embed_dim)(inputs)
x = layers.GlobalMaxPooling1D()(x)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(1, activation="sigmoid")(x)
model = keras.Model(inputs, outputs)
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.summary()
```

>[!Warning]
>**/keras_core/src/layers/layer.py:828: UserWarning: Layer 'global_max_pooling1d' (of type GlobalMaxPooling1D) was passed an input with a mask attached to it. However, this layer does not support masking and will therefore destroy the mask information. Downstream layers will not see the mask.**

<img width="819" alt="Screenshot 2023-10-04 at 10 35 55 AM" src="https://github.com/keras-team/keras/assets/11736571/a0de646f-3e75-4b3b-8a68-ee57c06c3612">

<br><br>

@fchollet do you think the bug is related to the same issue as pointed out in #18416 ? If you have any pointers, please let me know. I can fix this and send a PR

**Comment by AakashKumarNain** [comment]
I will check that right away.

**Comment by AakashKumarNain** [comment]
I checked it, and the output is same in both the cases, at least up to 5th decimal place

**Comment by qlzh727** [comment]
Seems that this warning is intended to notify user that the mask propagation is stopped due to the usage of certain layer within the model. I will close this, and feel free to reopen if you feel there is anything else need to be addressed here.

**Comment by jchwenger** [comment]
Hi there, I'm encountering this as well, when adapting the `TransformerEncoder` and `TransformerDecoder` stack found in *Deep Learning with Python* to work with the current API (for teaching).

Even if the warning is useful, often that comes with a way of making the code more specific so that the warning does not show up (the user has taken steps to confirm that's the desired behaviour). Is there such a thing in this case by any chance? Thanks in advance!
